cmd/config.go

```
@@ -0,0 +1,240 @@
+	return WriteFile(name, keyBytes, 0600)
+}
+
+// WriteFiles writes the files to disk
+func (cconf *CertificateConf) WriteFiles() (err error) {
+	if cconf.BundleCertificateFile == "" {
+		if err = WriteCertificates(cconf.CertificateFile, cconf.ClientCertificate); err != nil {
+			return
+		}
+		if err = WriteCertificates(cconf.RootCertificateFile, cconf.RootCertificate...); err != nil {
+			return
+		}
+	} else {
+		certs := append(cconf.RootCertificate, cconf.ClientCertificate)
+		if err = WriteCertificates(cconf.BundleCertificateFile, certs...); err != nil {
+			return
+		}
+	}
+	err = WritePrivateKey(cconf.KeyFile, cconf.ClientKey)
+	return
+}
+
+// RunCmd execs the cmd specified
+func (cconf *CertificateConf) RunCmd() (output string, err error) {
+	if cconf.Cmd != "" {
+		cmd := exec.Command("/bin/sh", "-c", cconf.Cmd) // #nosec
+		var boutput []byte
+		boutput, err = cmd.CombinedOutput()
+		output = string(boutput)
+	}
+	return
+}
+
+// DecodeValidityFuncsHookFunc is helper function for viper to parse our validityfunctions to a slice
+func DecodeValidityFuncsHookFunc() mapstructure.DecodeHookFunc {
+	return func(rf reflect.Type, rt reflect.Type, data interface{}) (interface{}, error) {
+		// test the data we got
+		if rf != reflect.TypeOf(map[interface{}]interface{}{}) {
+			return data, nil
+		}
+		// The slice of functions to contain results
+		var tests []func(*CertificateConf) error
+		// test the wanted type
+		if rt != reflect.TypeOf(tests) {
+			return data, nil
+		}
+		// raw repesentation of the map
+		raw := data.(map[interface{}]interface{})
+		if len(raw) == 0 {
+			return data, nil
+		}
+		for testKey, testValue := range raw {
+			ks, vs := testKey.(string), testValue.(string)
+			switch ks {
+			case "cname":
+				tests = append(tests, CheckCommonName())
+			case "cacname":
+				tests = append(tests, CheckCACommonName(vs))
+			case "ttl":
+				tests = append(tests, CheckTTL())
+			default:
+				// TODO: Unknown test key, should we return error?
```

Yes, I say we should return an error. If they ask for a test that doesn't exist, then something is wrong and will cause sadness.

---

Now it returns error